### ISSUE TITLE
android: Make it possible to build for a x86 target

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#18639]
I don't have a phone. I have to use an emulator. A arm emulator is a lot slower than a x86 emulator.

Can we make it so it's possible to build servo for i686-linux-android or x86_64-linux-android ?

**Comment by SimonSapin** [comment]
https://forge.rust-lang.org/platform-support.html lists support for `i686-linux-android`. Looks like the place to start is adding corresponding toolchain config in `handle_android_target` in `python/servo/command_base.py`.

**Comment by MortimerGoro** [comment]
I worked on adding Android armv7 and aarch64 architectures. Some required steps:

- Add the required x86 linker flags: https://github.com/servo/servo/blob/master/ports/servo/.cargo/config
- Add x86 configuration to https://github.com/servo/servo/blob/81044e001941ee3b4eef598125c327c73e993b5d/python/servo/command_base.py#L554 as @SimonSapin mentioned
- Adapt problematic cmake/makefiledependencies. Example of some changes required for armv7, aarch64: https://github.com/servo/servo/issues/11921
- Compile OpenSSL for x86. I already added some x86 flags when I updated OpenSSL script, it should work out of the box: https://github.com/servo/servo/pull/17099/files#diff-3196f1c36d8d0f70633f69275dfc6b4d
- For reference, I added x86 build types to the gradle project used to generate Servo APKs: https://github.com/servo/servo/blob/master/support/android/apk/app/build.gradle

**Comment by paulrouget** [comment]
@MortimerGoro I get this error:

```
   Compiling android_injected_glue v0.2.2
error[E0308]: mismatched types
   --> /Users/paul/.cargo/registry/src/github.com-1ecc6299db9ec823/android_injected_glue-0.2.2/src/lib.rs:245:88
    |
245 |                         if let Some(last_newline_pos) = buf.iter().rposition(|&c| c == b'\n') {
    |                                                                                        ^^^^^ expected i8, found u8

error[E0308]: mismatched types
   --> /Users/paul/.cargo/registry/src/github.com-1ecc6299db9ec823/android_injected_glue-0.2.2/src/lib.rs:246:53
    |
246 |                             buf[last_newline_pos] = b'\0';
    |                                                     ^^^^^ expected i8, found u8

error[E0308]: mismatched types
   --> /Users/paul/.cargo/registry/src/github.com-1ecc6299db9ec823/android_injected_glue-0.2.2/src/lib.rs:255:47
    |
255 |                                 buf[cursor] = b'\0';
    |                                               ^^^^^ expected i8, found u8

error: aborting due to 3 previous errors

error: Could not compile `android_injected_glue`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by SimonSapin** [comment]
This looks like `c_char` being `u8` or `i8` depending on the platform.

**Comment by paulrouget** [comment]
I think this has been fixed in tomaka's version of injected-glue, but not in mmatyas's. See:

https://github.com/tomaka/android-rs-glue/blob/master/cargo-apk/injected-glue/lib.rs#L270

vs

https://github.com/mmatyas/android-rs-injected-glue/blob/master/src/lib.rs#L245

**Comment by paulrouget** [comment]
https://github.com/mmatyas/android-rs-injected-glue/pull/3

**Comment by paulrouget** [comment]
I got it to work. The latest Android emulator is needed, and in the settings, ES 3.1 needs to be activated.

Then segfault. Still trying to figure that one out. (unrelated to #19058).

**Comment by paulrouget** [comment]
@MortimerGoro did the situation change?

**Comment by jdm** [comment]
This was fixed by #20472.
